go-types/filesource/filesource.go

```diff
@@ -0,0 +1,563 @@
+}
+
+type Listing interface {
+	// Next returns the next ObjectInfo, an io.EOF if no entries remain,
+	// or another encountered error.
+	Next() (ObjectInfo, error)
+}
+
+// ListingFunc adapts a listing closure to a Listing interface.
+type ListingFunc func() (ObjectInfo, error)
+
+func (f ListingFunc) Next() (ObjectInfo, error) { return f() }
+
+// ObjectInfo is returned by a Listing.
+type ObjectInfo struct {
+	Path            string
+	IsPrefix        bool
+	ContentSum      string
+	Size            int64
+	ContentType     string
+	ContentEncoding string
+	ModTime         time.Time
+}
+
+// ----------------------------------------------------------------
+//
+
+type connector struct {
+	config Config
+	store  Store
+}
+
+func newConnector(src Source, args airbyte.ConfigFile) (*connector, error) {
+	var cfg = src.NewConfig()
+
+	if err := args.ConfigFile.Parse(cfg); err != nil {
+		return nil, fmt.Errorf("parsing configuration: %w", err)
+	}
+
+	var store, err = src.Connect(context.Background(), cfg)
+	if err != nil {
+		return nil, fmt.Errorf("building filesystem: %w", err)
+	}
+
+	return &connector{
+		config: cfg,
+		store:  store,
+	}, nil
+}
+
+func (src Source) Main() {
+	var parserSpec, err = parser.GetSpec()
+	if err != nil {
+		panic(err)
+	}
+	var spec = airbyte.Spec{
+		SupportsIncremental:           true,
+		SupportedDestinationSyncModes: airbyte.AllDestinationSyncModes,
+		ConnectionSpecification:       src.ConfigSchema(parserSpec),
+	}
+	airbyte.RunMain(spec, src.Check, src.Discover, src.Read)
+}
+
+func (src Source) Check(args airbyte.CheckCmd) error {
+	var result = &airbyte.ConnectionStatus{
+		Status: airbyte.StatusSucceeded,
+	}
+
+	var _, err = newConnector(src, args.ConfigFile)
+	if err != nil {
+		result.Status = airbyte.StatusFailed
+		result.Message = err.Error()
+	}
+
+	return airbyte.NewStdoutEncoder().Encode(airbyte.Message{
+		Type:             airbyte.MessageTypeConnectionStatus,
+		ConnectionStatus: result,
+	})
+}
+
+func (src Source) Discover(args airbyte.DiscoverCmd) error {
+	var conn, err = newConnector(src, args.ConfigFile)
+	if err != nil {
+		return err
+	}
+	var ctx = context.Background()
+
+	// Breadth-first search.
+	var stack = []string{conn.config.DiscoverRoot()}
+	var streams []airbyte.Stream
+
+	for len(stack) != 0 && len(streams) < 10 {
+		var prefix = stack[0]
+
+		var listing, err = conn.store.List(ctx, Query{
+			Prefix:    prefix,
+			Recursive: false,
+		})
+		if err != nil {
+			return fmt.Errorf("starting listing %q: %w", prefix, err)
+		}
+
+		var hasObjects bool
+		for i := 0; i != discoverListLimit; i++ {
+			var entry, err = listing.Next()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return fmt.Errorf("during listing %q: %w", prefix, err)
+			} else if entry.IsPrefix {
+				stack = append(stack, entry.Path)
+			} else {
+				hasObjects = true
+			}
+		}
+
+		// Make streams of non-empty directories and the discovery root,
+		// even if the latter is empty.
```

Comment:
It seems a little weird for the stream name to be an empty string, although I don't see anything that explicitly disallows it. Just thinking out loud, we could use `"/"` as a sentinel value for the root. It might be better for users to type `stream: /` in their flow.yaml as opposed to `stream: ""`.

Reply:
it's actually `stream: "the-bucket:"`

Reply:
Ah, gotcha! That makes more sense. Though now I'm waffling a bit on the `:` vs `/` separating the bucket and prefix. It's looking like most other tools use a `/` to separate them, and `stream: the-bucket/` seems a little more intuitive than `stream: "the-bucket:"`. I hadn't considered before how it might look in the yaml when you want to capture everything from a bucket.

Reply:
Sounds good, I'll change back to '/' to separate bucket and key.